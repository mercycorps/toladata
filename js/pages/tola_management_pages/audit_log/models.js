import { observable, computed, action, runInAction } from "mobx"
import api from './api'

export class ProgramAuditLogStore {
    @observable program_id = null;
    @observable program_name = null;
    @observable log_rows = [];
    @observable fetching = false;
    @observable current_page = 0;

    @observable entries_count = 0;
    @observable total_pages = 0;
    @observable next_page = null;
    @observable previous_page = null;

    // UI state - track what history rows are expanded
    @observable expando_rows = new Set();

    constructor(program_id, program_name) {
        this.program_id = program_id;
        this.program_name = program_name;
        this.fetchProgramAuditLog();
    }

    @action
    fetchProgramAuditLog() {
        this.fetching = true

        api.fetchProgramAuditLogWithFilter(this.program_id, this.current_page + 1).then(results => {
            runInAction(() => {
                this.fetching = false;
                this.log_rows = results.logs;
                this.entries_count = results.total_entries;
                this.total_pages = results.total_pages;
                this.next_page = results.next_page;
                this.previous_page = results.previous_page;
            })
        })
    }

    @action
    changePage(page) {
        if(page.selected != this.current_page) {
            this.current_page = page.selected
            this.fetchProgramAuditLog()
        }
    }

    @action
    toggleRowExpando(row_id) {
        if (this.expando_rows.has(row_id)) {
            this.expando_rows.delete(row_id);
        } else {
            this.expando_rows.add(row_id);
        }
    }

    @action
    expandAllExpandos() {
        this.log_rows.forEach((row) => this.expando_rows.add(row.id));
    }

    @action
    collapsAllExpandos() {
        this.expando_rows.clear();
    }
}
