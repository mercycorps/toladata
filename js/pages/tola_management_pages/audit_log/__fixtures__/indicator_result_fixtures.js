export const auditLogIndicator = {
    level_display_ontology: "",
    leveltier_name: "Goal",
    name: "A test indicator",
    results_aware_number: "a"
};

export const auditLogResult = {
    id: "1",
    date: "2020-05-15"
}
