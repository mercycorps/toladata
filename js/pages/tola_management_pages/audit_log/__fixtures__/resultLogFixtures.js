export const resultLogChanged = {
    "name": "disaggregation_values",
    "pretty_name": "disaggregation_values",
    "prev": {
      "1": {
        "name": "Type One Value One",
        "value": "",
        "id": 1,
        "custom_sort": 1,
        "type": "Type of Leader"
      },
      "10": {
        "name": "Type One Value Two",
        "value": "",
        "id": 10,
        "custom_sort": 2,
        "type": "Type of Leader"
      },
      "5": {
        "name": "Type Two Value One",
        "value": "",
        "id": 5,
        "custom_sort": 8,
        "type": "Another Disaggregation"
      },
    },
    "new": {
      "1": {
        "name": "Type One Value One",
        "value": "2.00",
        "id": 1,
        "custom_sort": 2,
        "type": "Type of Leader"
      },
      "10": {
        "name": "Type One Value Two",
        "value": "23.00",
        "id": 10,
        "custom_sort": 1,
        "type": "Type of Leader"
      },
      "5": {
        "name": "Type Two Value One",
        "value": "18.00",
        "id": 5,
        "custom_sort": 8,
        "type": "Another Disaggregation"
      }
    }
};



export const resultLogDisaggs = {
    "1": {
        "name": "Value Two",
        "value": "",
        "id": 1,
        "custom_sort": 2,
        "type": "Type of Leader"
    },
    "10": {
        "name": "Value One",
        "value": "",
        "id": 10,
        "custom_sort": 1,
        "type": "Type of Leader"
    },
    "5": {
        "name": "Value Three",
        "value": "",
        "id": 5,
        "custom_sort": 3,
        "type": "Type of Leader"
    },
    "20": {
        "name": "Value Four",
        "value": "",
        "id": 20,
        "custom_sort": 4,
        "type": "Type of Leader"
    }
}
