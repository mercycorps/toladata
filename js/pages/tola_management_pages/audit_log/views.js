import React from 'react';
import { observer } from "mobx-react"
import Pagination from '../../../components/pagination'
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import { toJS } from "mobx"
import classNames from "classnames";

import LoadingSpinner from '../../../components/loading-spinner'

const emptyValue = "—";

export const DisaggregationDiffs = ({disagg_type, disagg_diffs}) => {
    disagg_diffs.sort( (a, b) => a.custom_sort - b.custom_sort);
    return <div><h4 className="disagg-type__title text-small">{gettext(disagg_type)}</h4>
        {disagg_diffs.map( diff => {
            const displayValue = ["", null, undefined].includes(diff.value)
                ? emptyValue
                : localizeNumber(normalizeNumber(diff.value));
            const displayClasses = classNames("change__field__value", {"empty-value": displayValue===emptyValue});
            return <div className="change__field" key={diff.id}><span className="change__field__name">{diff.name}:</span> <span className={displayClasses}>{displayValue}</span></div>
        })}
    </div>
};

export const ResultChangeset = ({data, name, pretty_name}) => {
    let displayValue = "";
    if (["", null, undefined].includes(data)) {
        displayValue = emptyValue;
    }
    else if (isNaN(data)) {
        displayValue = data;
    }
    else {
        displayValue = localizeNumber(data);
    }
    if (name === 'id') {
        return null
    } else if(name === 'Target_url') {
        return <div className="change__field"><strong className="change__field__name">{pretty_name}</strong>: {(data !== 'N/A' && data !== '')?<a href={displayValue} target="_blank">Link</a>:data}</div>
    } else if (name === 'disaggregation_values') {
        if (Object.entries(data).length) {
            let groupedDiffs = {};
            Object.values(data).forEach( entry => {
                const groupKey = entry.type || "__none__";
                if (entry.type in groupedDiffs) {
                    groupedDiffs[groupKey].push(entry);
                }
                else {
                    groupedDiffs[groupKey] = [entry];
                }
            });

            return <div className="changelog__change__targets">
                {Object.keys(groupedDiffs).sort().map( (typeName ) => {
                    return  <DisaggregationDiffs
                        key={typeName+'_diff'}
                        disagg_type={typeName === "__none__" ? "" : typeName } disagg_diffs={groupedDiffs[typeName]} />
                })}
            </div>
        } else {
            return null;
        }
    } else {
        const displayClasses = classNames("change__field__value", {"empty-value": displayValue===emptyValue});
        return <div className="change__field"><strong className="change__field__name">{pretty_name}</strong>: <span className={displayClasses}>{displayValue}</span></div>
    }
};

const ProgramDatesChangeset = ({data, name, pretty_name}) => {
    const displayValue = ["", null].includes(data)  ? "–" : data;
    return <p>{pretty_name}: {displayValue}</p>
}

const IndicatorChangeset = ({data, name, pretty_name, indicator}) => {
    if(name === 'targets') {
        return <div className="changelog__change__targets">
            <h4 className="text-small">{gettext('Targets changed')}</h4>
            {Object.entries(data).map(([id, target]) => {
                const displayValue = ["", null, undefined].includes(target.value) ? emptyValue : localizeNumber(target.value);
                const displayClasses = classNames({"empty-value": displayValue===emptyValue});
                return <div className="change__field" key={id}><strong className="change__field__name">{target.name}:</strong> <span className={displayClasses}>{displayValue}</span></div>
            })}
        </div>
    } else {
        let displayValue = "";
        if (["", null, undefined].includes(data)) {
            displayValue = emptyValue;
        }
        else if (isNaN(data)) {
            displayValue = data;
        }
        else {
            displayValue = localizeNumber(data);
        }
        const displayClasses = classNames({"empty-value": displayValue===emptyValue});
        return <div className="change__field">
            <strong className="change__field__name">
                { name === 'name' ?
                    <span>{gettext('Indicator')} {indicator.results_aware_number}: </span> : <span>{pretty_name}: </span>}
            </strong>
            <span className={displayClasses}>{displayValue}</span>
        </div>
    }
}

const ResultLevelChangeset = ({data, name, pretty_name, level}) => {
    const displayValue = ["", null, undefined].includes(data) ? emptyValue : data.toString();
    const displayClasses = displayValue === emptyValue ? "empty-value" : null;
    return <div className="change__field">
        { name !== 'name' ? <strong className="change__field__name">{pretty_name}: </strong>  : <strong><span className="field__level-tier">{level.tier} {level.display_ontology}: </span></strong> }
        <span className={displayClasses}>{displayValue}</span>
    </div>
}

class ChangesetEntry extends React.Component {
    renderType(type, data, name, pretty_name, indicator, level) {
        switch(type) {
            case 'indicator_changed':
            case 'indicator_created':
            case 'indicator_deleted':
                return <IndicatorChangeset data={data} name={name} pretty_name={pretty_name} indicator={indicator} level={level}/>
                break
            case 'result_changed':
            case 'result_created':
            case 'result_deleted':
                return <ResultChangeset data={data} name={name} pretty_name={pretty_name} />
                break
            case 'program_dates_changed':
                return <ProgramDatesChangeset data={data} name={name} pretty_name={pretty_name} />
                break
            case 'level_changed':
                return <ResultLevelChangeset
                    data={data} name={name} pretty_name={pretty_name} level={level} />
                break
        }
    }

    render() {
        const {data, type, name, pretty_name, indicator, level} = this.props;
        return this.renderType(type, data, name, pretty_name, indicator, level)
    }
}

const ExpandAllButton = observer(
    ({store}) => {
        return <button className="btn btn-medium text-action btn-sm"
                       onClick={() => store.expandAllExpandos()}
                       disabled={store.log_rows.length === store.expando_rows.size}>
            <i className="fas fa-plus-square"></i>
            {
                /* # Translators: button label to show the details of all rows in a list */}
            {gettext('Expand all')}
        </button>
    }
);

const CollapseAllButton = observer(
    ({store}) => {
        return <button className="btn btn-medium text-action btn-sm"
                       onClick={() => store.collapsAllExpandos()}
                       disabled={store.expando_rows.size === 0}>
            <i className="fas fa-minus-square"></i>
            {
                /* # Translators: button label to hide the details of all rows in a list */}
            {gettext('Collapse all')}
        </button>
    }
);

const IndicatorNameSpan = ({indicator}) => {
    if (! indicator) {
        return <span>{gettext('N/A')}</span>
    }

    if (indicator.results_aware_number) {
        return <span>
            <strong>{gettext('Indicator')} {indicator.results_aware_number}:</strong> {indicator.name}
        </span>
    } else {
        return <span>
            <strong>{gettext('Indicator')}:</strong> {indicator.name}
        </span>
    }
};

const ResultLevel = ({indicator, level}) => {
    if (level) {
        return `${level.tier} ${level.display_ontology}`;
    }

    if (indicator) {

        if (indicator.leveltier_name && indicator.level_display_ontology)
            return `${indicator.leveltier_name} ${indicator.level_display_ontology}`;
        else if (indicator.leveltier_name)
            return indicator.leveltier_name;
    }

    return <span>{gettext('N/A')}</span>
};

export const IndexView = observer(
    ({store}) => {
        return <div id="audit-log-index-view">
            <header className="page-title">
                <h1 className="page-title h2">
                    <a href={`/program/${store.program_id}/`}>{store.program_name}</a>: <span className="font-weight-normal text-muted text-nowrap">{gettext("Indicator change log")}&nbsp;<small><i className="fa fa-history" /></small></span>
                </h1>
            </header>

            <div className="admin-list__controls">
                <div className="controls__bulk-actions">
                    <div className="btn-group">
                        <ExpandAllButton store={store} />
                        <CollapseAllButton store={store} />
                    </div>
                </div>
                <div className="controls__buttons">
                    <a className="btn btn-secondary btn-sm" href={`/api/tola_management/program/${store.program_id}/export_audit_log`}>
                        <i className="fas fa-download" />
                        {gettext("Excel")}
                    </a>
                </div>
            </div>

            <div className="admin-list__table">
                <LoadingSpinner isLoading={store.fetching}>
                    <table className="table table-sm table-bordered bg-white text-small changelog">
                        <thead>
                            <tr>
                                <th className="text-nowrap">{gettext("Date and time")}</th>
                                <th className="text-nowrap">{gettext("Result level")}</th>
                                <th className="text-nowrap">{gettext("Indicator")}</th>
                                <th className="text-nowrap">{gettext("User")}</th>
                                <th className="text-nowrap">{gettext("Organization")}</th>
                                <th className="text-nowrap">{gettext("Change type")}</th>
                                <th className="text-nowrap">{gettext("Previous entry")}</th>
                                <th className="text-nowrap">{gettext("New entry")}</th>
                                <th className="text-nowrap">{gettext("Reason for change")}</th>
                            </tr>
                        </thead>
                        {store.log_rows.map(data => {
                                let is_expanded = store.expando_rows.has(data.id);
                                return <tbody key={data.id}>
                                <tr
                                    className={is_expanded ? 'changelog__entry__header is-expanded' : 'changelog__entry__header'}
                                    onClick={() => store.toggleRowExpando(data.id)}>
                                    <td className="text-action">
                                        <FontAwesomeIcon icon={is_expanded ? 'caret-down' : 'caret-right'} />
                                        &nbsp;{data.date} (UTC)
                                    </td>
                                    <td><ResultLevel indicator={data.indicator} level={data.level} /></td>
                                    <td>{<IndicatorNameSpan indicator={data.indicator} />}</td>
                                    <td>{data.user}</td>
                                    <td>{data.organization}</td>
                                    <td className="text-nowrap">{data.pretty_change_type}</td>
                                    <td></td>
                                    <td></td>
                                    <td></td>
                                </tr>
                                {is_expanded &&
                                <tr className="changelog__entry__row" key={data.id}>
                                    <td></td>
                                    <td></td>
                                    <td></td>
                                    <td></td>
                                    <td></td>
                                    <td></td>
                                    <td className="changelog__change--prev">
                                        {data.diff_list.map(changeset => {
                                            return <ChangesetEntry key={changeset.name} name={changeset.name}
                                                                   pretty_name={changeset.pretty_name}
                                                                   type={data.change_type} data={changeset.prev}
                                                                   indicator={data.indicator} level={data.level}/>
                                        })}
                                    </td>
                                    <td className="changelog__change--new">
                                        {data.diff_list.map(changeset => {
                                            return <ChangesetEntry key={changeset.name} name={changeset.name}
                                                                   pretty_name={changeset.pretty_name}
                                                                   type={data.change_type} data={changeset.new}
                                                                   indicator={data.indicator} level={data.level}/>
                                        })}
                                    </td>
                                    <td className="changelog__change--rationale">{data.rationale}</td>
                                </tr>
                                }
                                </tbody>
                            })}
                    </table>
                </LoadingSpinner>
                <div className="admin-list__metadata">
                    <div className="metadata__count text-muted text-small">{store.entries_count?`${store.entries_count} ${gettext("entries")}`:`--`}</div>
                    <div className="metadata__controls">
                        {store.total_pages &&
                         <Pagination
                             pageCount={store.total_pages}
                             initialPage={store.current_page}
                             onPageChange={page => store.changePage(page)} />
                        }
                    </div>
                </div>
            </div>
        </div>
    }
)
