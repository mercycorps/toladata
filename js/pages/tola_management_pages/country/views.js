import React from 'react'
import { observer } from "mobx-react"
import CheckboxedMultiSelect from 'components/checkboxed-multi-select'
import ManagementTable from 'components/management-table'
import Pagination from 'components/pagination'
import CountryEditor from './components/country_editor'
import EditCountryProfile from './components/edit_country_profile'
import EditDisaggregations from './components/edit_disaggregations'
import EditObjectives from './components/edit_objectives'
import LoadingSpinner from 'components/loading-spinner'
import FoldingSidebar from 'components/folding-sidebar'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import CountryHistory from "./components/country_history";

const CountryFilter = observer(({store, filterOptions}) => {
    return <div className="form-group react-multiselect-checkbox">
        <label htmlFor="countries_filter">{gettext("Countries")}</label>
        <CheckboxedMultiSelect
            value={store.filters.countries}
            options={filterOptions}
            onChange={(e) => store.changeFilter('countries', e)}
            placeholder={gettext("None Selected")}
            id="countries_filter" />
    </div>
})

const OrganizationFilter = observer(({store, filterOptions}) => {
    return <div className="form-group react-multiselect-checkbox">
        <label htmlFor="organizations_filter">{gettext("Organizations")}</label>
        <CheckboxedMultiSelect
            value={store.filters.organizations}
            options={filterOptions}
            onChange={(e) => store.changeFilter('organizations', e)}
            placeholder={gettext("None Selected")}
            id="organizations_filter" />
    </div>
})

const ProgramFilter = observer(({store, filterOptions}) => {
    return <div className="form-group react-multiselect-checkbox">
        <label htmlFor="programs-filter">{gettext("Programs")}</label>
        <CheckboxedMultiSelect
            value={store.filters.programs}
            options={filterOptions}
            onChange={(e) => store.changeFilter('programs', e)}
            placeholder={gettext("None Selected")}
            id="programs-filter" />
    </div>
})

export const IndexView = observer(
    ({store}) => {
        const countryFilterOptions = store.allCountries.map(country => {return {value: country.id, label: country.country}})
        const organizationFilterOptions = Object.entries(store.organizations).map(([id, org]) => ({value: org.id, label: org.name}))
        const programFilterOptions = Object.entries(store.allPrograms).map(([id, program]) => ({value: program.id, label: program.name}))
        return <div id="country-management-index-view" className="row">
            <FoldingSidebar>
                <div className="filter-section">
                    <OrganizationFilter store={store} filterOptions={organizationFilterOptions} />
                    <ProgramFilter store={store} filterOptions={programFilterOptions} />
                    <CountryFilter store={store} filterOptions={countryFilterOptions} />
                </div>
                <div className="filter-section filter-buttons">
                    <button className="btn btn-primary" onClick={() => store.applyFilters()}>{gettext("Apply")}</button>
                    <button className="btn btn-reset" onClick={() => store.clearFilters()}>{gettext("Reset")}</button>
                </div>
            </FoldingSidebar>
            <div className="col admin-list">
                <header className="page-title">
                    <h1>{gettext("Admin:")} <small>{gettext("Countries")}</small></h1>
                </header>
                <div className="admin-list__controls">
                    <div className="controls__bulk-actions"></div>
                    {store.is_superuser &&
                    <div className="controls__buttons">
                        <a href="#" tabIndex="0" className="btn btn-link btn-add" onClick={() => store.addCountry()}>
                            <i className="fas fa-plus-circle"/>{gettext("Add Country")}
                        </a>
                    </div>
                    }
                </div>
                <LoadingSpinner isLoading={store.fetching_main_listing || store.applying_bulk_updates }>
                    <div className="admin-list__table">
                        <ManagementTable
                            newData={store.new_country}
                            data={store.countries}
                            keyField="id"
                            HeaderRow={({Col, Row}) =>
                                <Row>
                                    <Col size="1"></Col>
                                    <Col size="60">{gettext("Country")}</Col>
                                    <Col>{gettext("Organizations")}</Col>
                                    <Col>{gettext("Programs")}</Col>
                                    <Col>{gettext("Users")}</Col>
                                </Row>
                            }
                            Row={({Col, Row, data}) =>
                            <Row
                                expanded={data.id == store.editing_target}
                                Expando={observer(({Wrapper}) =>
                                    <Wrapper>
                                        <CountryEditor
                                            notifyPaneChange={(new_pane) => store.onProfilePaneChange(new_pane)}
                                            active_pane={store.active_editor_pane}
                                            new={data.id == 'new'}
                                            ProfileSection={observer(() =>
                                                <LoadingSpinner isLoading={
                                                    store.fetching_editing_data ||
                                                    store.saving ||
                                                    store.fetching_editing_history
                                                }>
                                                    <EditCountryProfile
                                                        new={data.id == 'new'}
                                                        country_data={data}
                                                        organizationOptions={organizationFilterOptions}
                                                        onUpdate={(id, data) => store.updateCountry(id, data)}
                                                        onCreate={(new_country_data) => store.saveNewCountry(new_country_data)}
                                                        errors={store.editing_errors}
                                                        onIsDirtyChange={is_dirty => store.setActiveFormIsDirty(is_dirty)}
                                                    />
                                                </LoadingSpinner>
                                            )}
                                            StrategicObjectiveSection={observer(() =>
                                                <LoadingSpinner isLoading={
                                                    store.fetching_editing_data ||
                                                    store.saving ||
                                                    store.fetching_editing_history
                                                }>
                                                    <EditObjectives
                                                        country_id={data.id}
                                                        objectives={store.editing_objectives_data}
                                                        addObjective={() => store.addObjective()}
                                                        onUpdate={(id, data) => store.updateObjective(id, data)}
                                                        onCreate={(data) => store.createObjective(data)}
                                                        onDelete={(id) => store.deleteObjective(id)}
                                                        errors={store.editing_objectives_errors}
                                                        clearErrors={() => store.clearObjectiveEditingErrors()}
                                                        onIsDirtyChange={is_dirty => store.setActiveFormIsDirty(is_dirty)}
                                                    />
                                                </LoadingSpinner>
                                            )}
                                            DisaggregationSection={observer(() =>
                                                <LoadingSpinner isLoading={
                                                    store.fetching_editing_data ||
                                                    store.saving ||
                                                    store.fetching_editing_history
                                                }>
                                                    <EditDisaggregations
                                                        country_id={data.id}
                                                        countryName={data.country}
                                                        disaggregations={store.editing_disaggregations_data}
                                                        addDisaggregation={() => store.addDisaggregation()}
                                                        assignLabelErrors={store.assignDisaggregationLabelErrors}
                                                        onDelete={store.deleteDisaggregation.bind(store)}
                                                        onArchive={(id) => store.archiveDisaggregation(id)}
                                                        onUnarchive={(id) => store.unarchiveDisaggregation(id)}
                                                        onUpdate={(id, data) => store.updateDisaggregation(id, data)}
                                                        onCreate={(data) => store.createDisaggregation(data)}
                                                        errors={store.editing_disaggregations_errors}
                                                        clearErrors={() => store.clearDisaggregationEditingErrors()}
                                                        onIsDirtyChange={is_dirty => store.setActiveFormIsDirty(is_dirty)}
                                                    />
                                                </LoadingSpinner>
                                            )}
                                            HistorySection={observer(() =>
                                                <LoadingSpinner isLoading={
                                                    store.fetching_editing_data ||
                                                    store.saving ||
                                                    store.fetching_editing_history
                                                }>
                                                    <CountryHistory
                                                        store={store}
                                                        history={store.editing_history}
                                                    />
                                                </LoadingSpinner>
                                            )}
                                            fetchObjectives={(countryId) => store.fetchObjectives(countryId)}
                                        />
                                    </Wrapper>
                                )}>
                                    <Col size="0.2">
                                    </Col>
                                    <Col size="2" className="td--stretch">
                                        <div className="expando-toggle icon__clickable" onClick={() => store.toggleEditingTarget(data.id)} >
                                            <div className="expando-toggle__icon">
                                                <FontAwesomeIcon icon={(store.editing_target == data.id) ? 'caret-down' : 'caret-right'} />
                                            </div>
                                            <div className="expando-toggle__label">
                                                <i className="fas fa-globe"/>&nbsp;
                                                {data.country || "---"}
                                            </div>
                                        </div>
                                    </Col>
                                    <Col className="text-nowrap">
                                        { data.organizations.length ?
                                            <a href={`/tola_management/organization/?countries[]=${data.id}`}>
                                                <i className="fas fa-building"/>&nbsp;
                                                {data.organizations.length} {gettext("Organizations")}
                                            </a>
                                        : '---'}
                                    </Col>
                                    <Col className="text-nowrap">
                                        { data.programCount ?
                                            <a href={`/tola_management/program/?countries[]=${data.id}`}>
                                                <i className="fas fa-cubes"/>&nbsp;
                                                {data.programCount} {gettext("Programs")}
                                            </a>
                                        : "---"}
                                    </Col>
                                    <Col className="text-nowrap">
                                        { data.user_count ?
                                            <a href={`/tola_management/user/?countries[]=${data.id}`}>
                                                <i className="fas fa-users"/>&nbsp;
                                                {data.user_count} {gettext("Users")}
                                            </a>
                                        : '---'  }
                                    </Col>
                                </Row>
                            }
                        />
                    </div>
                </LoadingSpinner>
                <div className="admin-list__metadata">
                    <div className="metadata__count text-muted text-small">{store.country_count ? `${store.country_count} ${gettext("countries")}`:`---`}</div>
                    <div className="metadata__controls">
                        {store.total_pages &&
                         <Pagination
                            pageCount={store.total_pages}
                            initialPage={store.current_page}
                            onPageChange={page => store.changePage(page)} />
                        }
                    </div>
                </div>
            </div>
        </div>
    }
)
