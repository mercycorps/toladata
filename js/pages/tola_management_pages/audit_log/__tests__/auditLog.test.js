import React from 'react';
import { resultLogChanged, resultLogDisaggs } from '../__fixtures__/resultLogFixtures';
import { ProgramAuditLogStore } from "../models"
import { ResultChangeset, DisaggregationDiffs } from "../views"
import renderer from 'react-test-renderer';

window.localizeNumber = function (number) { number };
window.normalizeNumber = function (number) { number };

describe("Audit log", () => {

    it("orders disaggregations correctly", () => {
        let disaggList = Object.values(resultLogDisaggs);
        let component = renderer.create(
            <DisaggregationDiffs
                disagg_type={disaggList['1'].type}
                disagg_diffs={disaggList}
            />
        );
        let renderedLabels = [];
        for (let i=1; i<5; i++) {
            renderedLabels.push(component.toJSON().children[i].children[0].children[0]);
        }
        expect(renderedLabels).toEqual([ 'Value One', 'Value Two', 'Value Three', 'Value Four' ]);

        resultLogDisaggs["10"]["custom_sort"] = 4;
        resultLogDisaggs["20"]["custom_sort"] = 1;
        disaggList = Object.values(resultLogDisaggs);
        component = renderer.create(
            <DisaggregationDiffs
                disagg_type={disaggList['1'].type}
                disagg_diffs={disaggList}
            />
        );
        renderedLabels = [];
        for (let i=1; i<5; i++) {
            renderedLabels.push(component.toJSON().children[i].children[0].children[0]);
        }
        expect(renderedLabels).toEqual([ 'Value Four', 'Value Two', 'Value Three', 'Value One' ]);

    });

    it("orders disaggregations types correctly", () => {
        const component = renderer.create(
            <ResultChangeset
                name={resultLogChanged.name}
                pretty_name={resultLogChanged.pretty_name}
                data={resultLogChanged.new}
            />
        );
        let renderedLabels = [];
        for (let i=0; i<2; i++) {
            renderedLabels.push(component.toJSON().children[i].children[0].children[0]);
        }
        expect(renderedLabels).toEqual([ 'Another Disaggregation', 'Type of Leader' ]);
    })
});
