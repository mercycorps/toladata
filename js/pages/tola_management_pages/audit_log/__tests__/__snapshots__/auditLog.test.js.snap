// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit log displays N/A when the baseline not-applicable checkbox has been selected 1`] = `
<div
  className="change__field"
>
  <strong
    className="change__field__name"
  >
    <span>
      Baseline
      : 
    </span>
  </strong>
  <span
    className=""
  >
    N/A
  </span>
</div>
`;

exports[`Audit log displays an em-dash for the empty prev entry baseline when type is indicator created 1`] = `
<div
  className="change__field"
>
  <strong
    className="change__field__name"
  >
    <span>
      Baseline
      : 
    </span>
  </strong>
  <span
    className="empty-value"
  >
    â€”
  </span>
</div>
`;

exports[`Audit log displays the baseline number when a baseline has been entered 1`] = `
<div
  className="change__field"
>
  <strong
    className="change__field__name"
  >
    <span>
      Baseline
      : 
    </span>
  </strong>
  <span
    className=""
  >
    4
  </span>
</div>
`;

exports[`Audit log displays two components in indicator column - one for indicator and one for result 1`] = `
Array [
  <span>
    <strong>
      Indicator
       
      a
      :
    </strong>
     
    A test indicator
  </span>,
  <p
    className="mt-2"
  >
    <strong>
      Result date:
    </strong>
     
    2020-05-15
  </p>,
]
`;
